expression/util.go

```diff
@@ -288,46 +290,55 @@ var symmetricOp = map[opcode.Op]opcode.Op{
 	opcode.NullEQ: opcode.NullEQ,
 }
 
-func doPushDownNot(ctx sessionctx.Context, exprs []Expression, not bool) []Expression {
+func doPushDownNot(ctx sessionctx.Context, exprs []Expression, not bool) ([]Expression, bool) {
 	newExprs := make([]Expression, 0, len(exprs))
+	flag := false
 	for _, expr := range exprs {
-		newExprs = append(newExprs, PushDownNot(ctx, expr, not))
+		newExpr, changed := PushDownNot(ctx, expr, not)
+		flag = changed || flag
+		newExprs = append(newExprs, newExpr)
 	}
-	return newExprs
+	return newExprs, flag
 }
 
 // PushDownNot pushes the `not` function down to the expression's arguments.
-func PushDownNot(ctx sessionctx.Context, expr Expression, not bool) Expression {
+func PushDownNot(ctx sessionctx.Context, expr Expression, not bool) (Expression, bool) {
```

Comment:
Can we rewrite it like `FoldConstant` that only returns expression, but `foldConstant` would return other values, it could save many unnecessary changes.

Reply:
Not very easy for now.
`FoldConstant` is called everytime when the expr is built so we don't need to traverse the whole tree.
But `PushDownNot` is not, so we know the expr_1 needs to be generated as a new expr only when the subtree is checked. Thus, we need a extra returned value to tell us the status of the subtree.
If we want to ingore it. We need to move it inside the `NewFunction`.

Reply:
Oh, I mean we can rewrite it like
```
func PushDownNot(...) Expression
{
   expr, _ := pushDownNot(...)
   return expr
}
```
so people that uses `PushDownNot` do not need to care about the useless value.

Reply:
+1

Reply:
fixed